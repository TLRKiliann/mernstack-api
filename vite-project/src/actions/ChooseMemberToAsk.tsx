import React from 'react'
import { UserType } from '../models/usertype'
import { ComputerType } from '../models/computerType'

interface ChooseMemberToAskProps {
  user: UserType
  computer: ComputerType
  handleAskUserPrivate: (e: React.MouseEvent<HTMLButtonElement>, id: number) => void
  addUserById: (e: React.MouseMove<HTMLButtonElement>, id: number) => void
}

const ChooseMemberToAsk: React.FC = (props: ChooseMemberToAskProps) => {
  return (
    <span 
      data-testid="chooseMemberId"
      className="span--saloon">
      {props.user?.mainroom === props.computer?.title ? (
        <div className="user--imgsaloon">
          <img
            width="100%"
            height="100%"
            src={props.user?.img} 
            className="img--saloon"
          />
          <p className="useronline">
          {props.user?.firstName} {props.user?.isConnected ? (
            <span
              className="useronline--iconconnectected"
            >
              ✔
            </span>
            ):(
            <span
              className="useronline--icondeconnectected"
            >
              ❌
            </span>
            ) 
          }
          </p>
          <span
            onClick={() => props.handleAskUserPrivate(props.user?.id)}
            className="askprivate--service"
            title="Invite to private chat"
          >
            ✉
          </span>

          <span key={props.user?.id} className="lastspan--online">
            <button
              onClick={() => props.addUserById(props.user?.id)}
              className="btn--lastspanonline"
              title="Add as Your Friend"
            >
              +
            </button>
          </span>

        </div>
        ) : null
      }
    </span>
  )
}

export default ChooseMemberToAsk