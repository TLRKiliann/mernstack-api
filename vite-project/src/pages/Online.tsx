import React, { useState, useEffect } from 'react'
import { useNavigate } from 'react-router-dom'
import { useAuthLogin } from '../context/AuthProvider'
import { db_users } from '../models/db_users'
import { UserType } from '../models/usertype'
import { db_computers } from '../models/db_computers'
import { computerType } from '../models/computerType'
import AskMessageBox from '../components/AskMessageBox'
import ResponsePrivateMsg from '../components/ResponsePrivateMsg'
import '../stylePages/Online.scss'

type DataAllType = {
  db_users: UserType[]
  db_computers: computerType[]
}

type Field = {
  value?: sting
  label?: string
}

type Form = {
  invite: Field
}

const options: Field[] = [
  {
    label: "Private",
    value: "Private",
  },
  {
    label: "Rdv",
    value: "Rdv",
  },
  {
    label: "Briefing",
    value: "Briefing",
  }
]

const Online: React.FC = () => {

  const Navigate = useNavigate()

  const [form, setForm] = useState<Form>({
    invite: {value: 'Private'}
  })

  const [users, setUsers] = useState<Array<DataAllType>>([])
  const [computers, setComputers] = useState<Array<computerType>>([])
  const [group, setGroup] = useState<Array<string | number | boolean>>([])
  const [catchById, setCatchById] = useState<Array<UserType>>([])
  
  //console.log(catchById, "catchById")

  const [switchAsk, setSwitchAsk] = useState<boolean>(false)
  const [switchResponse, setSwitchResponse] = useState<boolean>(false)

  const { setOtherUser, setTweekGroup } = useAuthLogin();

  useEffect(() => {
    setUsers(db_users)
    setComputers(db_computers)
  }, [])

  const addUserById = (id: number) => {
    const userTodAdd = users.find(user => user.id === id)
    const verifyAlreadyExist = group.find(g => g.id === id)
    if (verifyAlreadyExist === userTodAdd) {
      console.log("This user already exists")
      setGroup(group.filter(g => g.id !== id))
    } else {
      console.log("Ok, new user accepted")
      setGroup([...group, userTodAdd])
      setTweekGroup([...group, userTodAdd])
      alert(`User ${userTodAdd.firstName} added to your group
        (look at your Profile !)`)
    }
  }

  const handleInviteChoice = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const fieldName: string = e.target.name;
    const fieldValue: string | number  = e.target.value;
    const newField: Field = {[fieldName]: {value: fieldValue}}

    setForm({...form, ...newField});
  }

  const handleAskUserPrivate = (id: number) => {
    const catchUser = users?.find(user => user.id === id)
    setCatchById(catchUser)
    setSwitchAsk(!switchAsk)
  }

  const handleInvitation = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault()
    //console.log(catchById, "catchUser")
    setOtherUser(catchById)
    setSwitchAsk(!switchAsk)
    setSwitchResponse(!switchResponse)
    //Navigate('/computerroom/privatemessage')
    
    /*const sendMessage = users?.find(user => user === catchById)
    console.log(sendMessage, "sendMessage")*/

    /*serviceRouting
      .getAllMembers(sendMessage, id)
      .then(initialData => {
        setUsers(initialData)
      })
      .catch((error) => {
        console.log("error", error)
      })
    */
  }

  const handleClose = () => {
    setSwitchAsk(false)
  }

  const handleCloseResponse = () => {
    setSwitchResponse(false)
  }

  const handleCloseConfirm = () => {
    setConfirmRequest(false)
  }

  return(
    <div className="saloon--byusers">

      <div className="title--services">
        <h1>Rooms</h1>
      </div>

      <div className="divsection--saloon">

      {switchAsk &&
        <AskMessageBox
          catchById={catchById}
          form={form.invite.value}
          options={options}
          handleInviteChoice={(e) => handleInviteChoice(e)}
          handleInvitation={(e) => handleInvitation(e)}
          handleClose={handleClose}
        />
      }

      {switchResponse && 
        <ResponsePrivateMsg
          form={form.invite.value}
          catchById={catchById}
          handleCloseResponse={handleCloseResponse}
        />
      }

      {Object.values(computers)?.slice(0, 12).map(computer =>
        <section key={computer.id} className="section--saloon">
          <h3 className="titlebyroom">{computer.title}</h3>

          {Object.values(users)?.map(user =>
            <span key={user.id} className="span--saloon">
              {user.mainroom === computer.title ? (
                <div className="user--imgsaloon">
                  <img
                    width="40px"
                    height="40px"
                    src={user.img} 
                    className="img--saloon"
                  />
                  <p style={{fontSize: '16px'}}>
                  {user.firstName} {user.isConnected ? (
                    <span
                      style={{
                        color: 'lightgreen',
                        marginLeft: "5px"
                      }}
                    >
                      ✔
                    </span>
                    ):(
                    <span
                      style={{
                        color: 'red',
                        marginLeft: "5px",
                        fontSize: "18px"
                      }}
                    >
                      ❌
                    </span>
                    ) 
                  }
                  </p>
                  <span
                    onClick={() => handleAskUserPrivate(user.id)}
                    className="askprivate--service"
                    title="Invite to private chat"
                  >
                    ✉
                  </span>

                  <span key={user.id} className="lastspan--online">
                    <button
                      onClick={() => addUserById(user.id)}
                      className="btn--lastspanonline"
                      title="Add as Your Friend"
                    >
                      +
                    </button>
                  </span>

                </div>
                ) : null
              }
            </span>
            )}
        </section>
      )}
      </div>
    </div>
  )
}

export default Online