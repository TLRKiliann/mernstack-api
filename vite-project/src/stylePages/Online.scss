.saloon--byusers {
  position: absolute;
  top: 0px;
  right: 0px;
  left: 0px;
  min-height: 100vh;
  background: linear-gradient(180deg, 
    purple, orange);

  .title--services {
    margin-top: 80px;
    margin-left: 20px;
    font-family: "Junction", sans-serif;
    text-shadow: 2px 1px 1px #000;
    color: lightgrey;
  }

  .divsection--saloon {
    width: 80%;
    margin: auto;
    display: grid;
    grid-column-gap: 20px;
    grid-row-gap: 20px;
    grid-template-columns: 1Fr 1Fr 1Fr 1Fr;
    grid-template-rows: 1Fr 1Fr 1Fr;
    margin-bottom: 40px;

    .titlebyroom {
      margin: auto;
      padding: 7px 0px;
      font-family: "Junction";
      text-align: center;
      background: purple;
      border-bottom: 1px solid slategray;
      border-radius: 15px 15px 0px 0px;
      color: lightgrey;
    }

    .section--saloon {
      background: #383838;
      border: none;
      border-radius: 15px;
      box-shadow: 0 0 4px #000;

      .span--saloon {
        display: flex;
        background: #242424;
        font-family: "DejaVu Sans";
        font-size: 18px;
        color: #50bcfe;
      }

      .user--imgsaloon {
        width: 100%;
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: flex-start;
        border-bottom: 1px solid slategray;
      }

      .img--saloon {
        margin: 0px 10px;
        border-radius: 50% 50%;
        object-fit: cover;
      }

      .askprivate--service {
        position: absolute;
        width: 15%;
        display: flex;
        justify-content: right;

        font-size: 22px;
        cursor: pointer;
        color: orange;
        z-index: 3;

        &:hover {
          color: cyan;
        }
      }

      .lastspan--service {
        position: absolute;
        width: 18%;  
        display: flex;
        align-items: center;

        justify-content: right;
        font-size: 14px;
        font-family: "DejaVu Sans";
        font-weight: thin;
        color: orange;

        .btn--lastspanservice {
          margin-left: 10px;
          padding: 3px 10px;
          cursor: pointer;
          background: #606060;
          border-radius: 50% 50%;
          font-size: 1.0rem;
          font-weight: bold;
          color: #50bcfe;

          &:hover {
            background: #707070;
            color: lightgreen;
          }
        }
      }
    }
  }
}