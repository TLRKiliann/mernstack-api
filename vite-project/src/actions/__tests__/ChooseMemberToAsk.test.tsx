import React from "react"
import { screen, render, fireEvent } from '@testing-library/react'
import { expect, vi } from 'vitest'
import { create } from 'react-test-renderer'
import '@testing-library/jest-dom'
import ChooseMemberToAsk from "../ChooseMemberToAsk"
import Online from '../../pages/Online'

test("mocking ChooseMemberToAsk", () => {
  vi.mock('../ChooseMemberToAsk.tsx', () => ({
    ChooseMemberToAsk: vi.fn()
  }))
})


const Span = ({onClick, children}) => (
<span onClick={onClick} data-testid="ask">{children}</span>
)
test('calls onClick prop when clicked handleAskUserPrivate', () => {
  const user = {id: 1}
  const handleAskUserPrivate = vi.fn(user => user)
  render(
    <Span onClick={handleAskUserPrivate} data-testid="ask">
      ✉
    </Span>
  )
  fireEvent.click(screen.getByText<HTMLSpanElement>(/✉/))
  expect(handleAskUserPrivate).toHaveBeenCalledTimes(1)
  const valAsk = screen.getByText(/✉/)
  expect(valAsk).toBeInTheDocument()
})

const Button = ({onClick, children}) => (
<button onClick={onClick} data-testid="add">{children}</button>
)
test('calls onClick prop when clicked addUserById', () => {
  const user = {id: 2}
  const addUserById = vi.fn(user => user)
  render(
    <button onClick={addUserById} data-testid="add">
      +
    </button>
  )
  fireEvent.click(screen.getByText<HTMLButtonElement>("+"))
  expect(addUserById).toHaveBeenCalledTimes(1)
  const valAsk = screen.getByText("+")
  expect(valAsk).toBeInTheDocument()
})

//Parent component to test child component
test("test child component", () => {
  const { getByText } = render(
    <Online />
  )
  expect(getByText(/Rooms/)).toBeInTheDocument()
})