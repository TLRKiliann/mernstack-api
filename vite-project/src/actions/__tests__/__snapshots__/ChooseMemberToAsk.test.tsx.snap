// Vitest Snapshot v1

exports[`ChooseMember MatchSnapShot 1`] = `
<span
  className="span--saloon"
  data-testid="chooseMemberId"
>
  <div
    className="user--imgsaloon"
  >
    <img
      className="img--saloon"
      height="100%"
      width="100%"
    />
    <p
      className="useronline"
    >
       
      <span
        className="useronline--icondeconnectected"
      >
        ❌
      </span>
    </p>
    <span
      className="askprivate--service"
      data-testid="ask"
      onClick={[Function]}
      title="Invite to private chat"
    >
      ✉
    </span>
    <span
      className="lastspan--online"
    >
      <button
        className="btn--lastspanonline"
        data-testid="add"
        onClick={[Function]}
        title="Add as Your Friend"
      >
        +
      </button>
    </span>
  </div>
</span>
`;
